### PR TITLE
add files

### PR DESCRIPTION
**Pull request by SvetlanaYes**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import NN
+
+
+def main():
+    if len(sys.argv) != 2:
+        print("[config.json]")
+        return
+    parameters = u.load_file(sys.argv[1])
+    train_file, tokenization_type, language_model, embedding = u.get_embedding_parameters(parameters)
+    X_train, y_train = embd.get_x_y(tokenization_type, train_file, language_model, embedding)
```
Comment (tsolakghukasyan) [inline main.py:17]:
```suggestion
    X_train, y_train = embd.get_x_y(**parameters["embeddings"])
```

And then modify `embd.get_x_y` to accept `**kwargs` as parameters.

Similarly, you can change `model.train`. This way, the code will be more concise and `get_...` functions in utils will not be necessary.

File: embedding.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+def get_x_y(tokenization_type, file, model_name, embeddings):
+    model = SentenceTransformer(model_name)
+    X_train = []
+    y_train = []
+    pattern = "<EOP>"
+    with open(file, encoding='utf-8') as f:
+        csvreader = csv.reader(f, delimiter='\t')
+        for j, row in enumerate(csvreader):
+            title = row[1]
```
Comment (tsolakghukasyan) [inline embedding.py:26]: Instead of `row[1]`, `row[2]`, ... `row[5]` it's better to unwrap row on line 25 (`for j, (title, ...) in enumerate(csvreader):`). That way the code will be more readable.

File: config.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    },
+  "model":
+    {
+      "NN":  "LogisticRegression",
+      "criterion": "BCELoss",
+      "optimizer": "SGD"
+    },
+  "save_model":
+    {
+      "save_model_to": "log_reg_paragraph_tokenization.pkl"
```
Comment (tsolakghukasyan) [inline config.json:24]: For model names use unique IDs, and then keep a table that maps model ID to its config.

File: embedding.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    if embd_type == "subtract":
+        x = torch.sub(text, title)
+    if i in incongruent_paragraphs:
+        y = 1
+    else:
+        y = 0
+    return x, y
+
+
+def get_x_y(tokenization_type, file, model_name, embeddings):
```
Comment (tsolakghukasyan) [inline embedding.py:18]: The function is too big. Decompose into smaller functions.

File: embedding.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import csv
+from sentence_transformers import SentenceTransformer
+import torch
+
+
+def combine_2_embeddings(embd_type, text, title, i, incongruent_paragraphs):
+    if embd_type == "concatenate":
+        x = torch.cat(text, title)
+    if embd_type == "subtract":
+        x = torch.sub(text, title)
```
Comment (tsolakghukasyan) [inline embedding.py:10]: If `emb_type` is neither `concatenate` nor `subtract`, the variable x won't be initialized. The function should raise `ValueError` with appropriate message, when `embd_type` is not specified correctly.

File: embedding.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    if i in incongruent_paragraphs:
+        y = 1
+    else:
+        y = 0
```
Comment (tsolakghukasyan) [inline embedding.py:14]:
```
    if i in incongruent_paragraphs:
        y = 1
    else:
        y = 0
```
This part is not strictly related to the way `x` is computed, so it should be a separate function. Also, it can be done using one line (e.g. `y = int(i in incongruent_paragraphs)`).

File: NN.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        y_train = torch.Tensor(y_train)
+        if self.model_name == "LogisticRegression":
+            model = LogisticRegression(input_dim, output_dim)
+        if self.criterion == "BCELoss":
+            criterion = torch.nn.BCELoss()
+        if self.optimizer == "SGD":
+            optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
+
+        losses = []
+        Iterations = []
```
Comment (tsolakghukasyan) [inline NN.py:39]: variable names should be lowercase.

File: NN.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            loss = criterion(torch.squeeze(outputs), labels)
+            loss.backward()
+            optimizer.step()
+            iter += 1
+            if iter % 10000 == 0:
+                total = 0
+                correct = 0
+                total += y_train.size(0)
+                correct += np.sum(torch.squeeze(outputs).round().detach().numpy() == y_train.detach().numpy())
+                accuracy = 100 * correct / total
```
Comment (tsolakghukasyan) [inline NN.py:54]: Accuracy computation should be a separate function.

File: NN.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        if self.optimizer == "SGD":
+            optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
+
+        losses = []
+        Iterations = []
+        iter = 0
+        for epoch in tqdm(range(int(epochs)), desc='Training Epochs'):
+            labels = y_train
+            optimizer.zero_grad()
+            outputs = model(X_train)
```
Comment (tsolakghukasyan) [inline NN.py:44]: You are currently implementing batch gradient descent, which is slow and has high memory requirements. You should split the train dataset into minibatches and use minibatch gradient descent for training. The size of minibatch should be a hyperparameter in `config.json` (you can name it `batch_size`).

File: NN.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    def train(self, X_train, y_train, epochs, input_dim, output_dim, learning_rate):
+        X_train = torch.Tensor(X_train)
+        y_train = torch.Tensor(y_train)
+        if self.model_name == "LogisticRegression":
+            model = LogisticRegression(input_dim, output_dim)
+        if self.criterion == "BCELoss":
+            criterion = torch.nn.BCELoss()
+        if self.optimizer == "SGD":
+            optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
```
Comment (tsolakghukasyan) [inline NN.py:36]: Current implementation only supports using SGD optimizer. There are more advanced optimizer (e.g. Adam), which can yield better results. Change the code to allow using other optimizers.

File: embedding.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+                    sentences = element.split("<EOS>")
+                    for sentence in sentences:
+                        sentence_embedding = model.encode(sentence, convert_to_tensor=True)
+                        x.append(combine_2_embeddings(embeddings, sentence_embedding, title_embedding))
+                        y.append(int(i in incongruent_paragraphs))
+
+                else:
+                    body_embedding = model.encode(element, convert_to_tensor=True)
+                    x.append(combine_2_embeddings(embeddings, body_embedding, title_embedding))
+                    y.append(int(i in incongruent_paragraphs))
```
Comment (tsolakghukasyan) [inline embedding.py:51]: Lines 44-46 and 49-51 contain code duplication. 

```suggestion
                if tokenization_type == "<EOS>":
                    sentences = element.split("<EOS>")
                else:
                    sentences = element
                    
                for sentence in sentences:
                     sentence_embedding = model.encode(sentence, convert_to_tensor=True)
                     x.append(combine_2_embeddings(embeddings, sentence_embedding, title_embedding))
                     y.append(int(i in incongruent_paragraphs))
```

File: embedding.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        x = torch.sub(text, title)
+    else:
+        raise ValueError(f"Invalid embedding type: {embd_type}. Should be concatenate or subtract")
+    return x
+
+
+def get_incongruent_paragraphs(label, incongruent_sentences):
+    incongruent_paragraphs = []
+    if label == 1:
+        if len(incongruent_sentences) > 2:
```
Comment (tsolakghukasyan) [inline embedding.py:19]:
```suggestion
    if label == 1 and len(incongruent_sentences) > 2:
```

File: embedding.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            incongruent_sentences = incongruent_sentences[1:-1].split(', ')
+            for el in incongruent_sentences:
+                incongruent_paragraphs.append(int(el))
```
Comment (tsolakghukasyan) [inline embedding.py:22]:
```suggestion
            incongruent_paragraphs = list(map(lambda el: int(el), incongruent_sentences[1:-1].split(', '))
```

File: NN.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        optimizer = u.optimizers[self.optimizer](model.parameters(), lr=learning_rate)
+        criterion = u.criterion[self.criterion]()
+        losses = []
+        losses_test = []
+        iterations = []
+        iter = 0
+        n_batches = int(len(X_train) / batch_size)
+        for epoch in tqdm(range(int(epochs)), desc='Training Epochs'):
+            for batch in range(n_batches):
+                batch_X, batch_y = X_train[batch * batch_size:(batch + 1) * batch_size ], y_train[batch * batch_size:(batch + 1) * batch_size ]
```
Comment (tsolakghukasyan) [inline NN.py:59]: This kind of batching doesn't perform shuffling of examples before each epoch, which is essential for minibatch sgd.
